src/pydicom/multival.py

```diff
@@ -8,131 +8,162 @@
+    def __setitem__(self, idx: int, val: T) -> None:
         pass  # pragma: no cover
 
     @overload
-    def __setitem__(self, idx: slice, val: Iterable[_T]) -> None:
+    def __setitem__(self, idx: slice, val: Iterable[T]) -> None:
         pass  # pragma: no cover
 
-    def __setitem__(  # type: ignore[misc]
-        self, idx: int | slice, val: _T | Iterable[_T]
-    ) -> None:
-        """Set an item of the list, making sure it is of the right VR type"""
-        if isinstance(idx, slice):
-            val = cast(Iterable[_T], val)
-            out = [self.type_constructor(v) for v in val]
-            self._list.__setitem__(idx, out)
+    def __setitem__(self, index: slice | int, val: Iterable[T] | T) -> None:
+        """Add item(s) at `index`."""
+        if isinstance(index, slice):
+            val = cast(Iterable[T], val)
+            self._list.__setitem__(index, [self._validate(item) for item in val])
         else:
-            val = cast(_T, val)
-            self._list.__setitem__(idx, self.type_constructor(val))
+            val = cast(T, val)
+            self._list.__setitem__(index, self._validate(val))
+
+    def _validate(self, item: Any) -> T:
+        """Return items that have been validated as being of the expected type"""
+        raise NotImplementedError(
+            f"'{type(self).__name__}._validate()' must be implemented"
+        )
+
+
+class MultiValue(ConstrainedList):
+    """Class to hold any multi-valued DICOM value, or any list of items that
+    are all of the same type.
+
+    This class enforces that any items added to the list are of the correct
+    type, by calling the constructor on any items that are added. Therefore,
+    the constructor must behave nicely if passed an object that is already its
+    type. The constructor should raise :class:`TypeError` if the item cannot be
+    converted.
+
+    Note, however, that DS and IS types can be a blank string ``''`` rather
+    than an instance of their classes.
+    """
+
+    def __init__(
+        self,
+        type_constructor: Callable[[Any], T],
+        iterable: Iterable[Any],
+        validation_mode: int | None = None,
+    ) -> None:
+        """Create a new :class:`MultiValue` from an iterable and ensure each
+        item in the :class:`MultiValue` has the same type.
+
+        Parameters
+        ----------
+        type_constructor : callable
+            A constructor for the required type for all items. Could be
+            the class, or a factory function. Takes a single parameter and
+            returns the input as the desired type (or raises an appropriate
+            exception).
+        iterable : iterable
+            An iterable (e.g. :class:`list`, :class:`tuple`) of items to
+            initialize the :class:`MultiValue` list. Each item in the iterable
+            is passed to `type_constructor` and the returned value added to
+            the :class:`MultiValue`.
+        """
```

Comment:
`validation_mode` parameter doc missing here.  Also, `type_constructor` can take two parameters, as `validation_mode` passed to it for DS and IS.

Reply:
Hmm, the logic around `validation_mode` and the `DSfloat/DSdecimal/ISfloat` behaviour is not particularly straight-forward. Is it OK if I submit a separate PR cleaning it all up?

For instance, `validation_mode` actually only applies to those three classes, trying to use it with anything will break things. `MultiString` also passes it when splitting up strings, but nothing that calls `MultiString` actually uses validation so it all works out. It seems like it'd be simpler to just remove the `validation_mode` and kick the `DSfloat/DSdecimal/ISfloat` behaviour back up to `DataElement._convert`

Reply:
Yes, that would make both `_constructor` and `_valmode` obsolete and would make `MultiValue` a more generic and lean class. Probably better.

Reply:
> Is it OK if I submit a separate PR cleaning it all up?

Yes, that sounds good.